### ISSUE TITLE
Hook timing of `edit_form_advanced`

### ISSUE THREAD

**Issue by moorscode** [WordPress/gutenberg#10929]
**Describe the bug**

In Yoast SEO Premium we use the `edit_form_advanced` hook to output a hidden input field.
On a Gutenberg page, this hook is being triggered at initialization.

This caused us a very strange bug, where the metabox under the content has very unwanted behaviour.

This used to only be the case for IE, but since Gutenberg 4.1, this also affects Chrome (Firefox can handle it).

The problem is that the hidden field (which is output twice, for some odd reason), is being placed before the opening `<html` tag of the document.

As this behaviour is pretty obscure, I'm most certain that other plugins also use `echo` or print to the document in this hook, resulting in unwanted rending of the page.

As mentioned in https://github.com/WordPress/gutenberg/issues/1352:
> `edit_form_advanced` - Fires after 'normal' context meta boxes have been output for all post types other than 'page'.

**To Reproduce**

Add a hook implementation:
```php
add_action( 'edit_form_advanced', function() {
  echo '<input type="hidden" name="some_name" value="something"/>';
}, 10, 1 );
```

This will create an invalid HTML document structure, which results in unreliable results in different browsers.

**Expected behavior**
I expect this hook to be called at the location where the content div has been rendered, not before any output is sent.

**Screenshots**
![weird-behaviour](https://user-images.githubusercontent.com/2005352/47343094-a3bfbd00-d6a5-11e8-8b1c-628eea52be57.gif)

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser Chrome, IE
 - Version latest

**Smartphone (please complete the following information):**
Untested

**Additional context**
Gutenberg 4.1-RC2

**Comment by azaozz** [comment]
Uh, that seems a "classic" example of doing-it-wrong? :)
The filters on the TinyMCE init array are meant to filter that array, not to trigger output. For outputting extra HTML together the editor's HTML, most appropriate would be to use `'the_editor'` filter which is meant for that. You can safely prepend or append any HTML there.

As that is already merged to core, opened https://core.trac.wordpress.org/ticket/45150 as follow up.

**Comment by swissspidy** [comment]
`edit_form_advanced` is not a TinyMCE related filter though? 🤔 

According to https://wpdirectory.net/search/01CTG7DNKS5FTNBZP6VY01EB6F many plugins print some custom HTML in `edit_form_advanced`, most notably https://wordpress.org/plugins/disable-comments/.

**Comment by 1000camels** [comment]
do_action( 'edit_form_advanced', $post ); is also in wp-admin/edit-form-advanced.php, where before it calls do_meta_boxes(null, 'normal', $post), and after it calls do_meta_boxes(null, 'advanced', $post), both of which output markup.
The other place that edit_form_advanced is called is in wp-includes/class-wp-embed.php, which also outputs markup.

So it sounds to me like gutenberg_collect_meta_box_data() should be calling some other action.

**Comment by danielbachhuber** [comment]
@moorscode Was this positively or negatively impacted by #10660?

**Comment by moorscode** [comment]
@danielbachhuber negatively, it was not as broken before (there were strange visual glitches before).

**Comment by danielbachhuber** [comment]
@moorscode Ok, thanks.

I think we need to keep #10660 in, because there were more significant issues caused by the prior implementation.

For this particular issue, one option to consider is:

```php
ob_start();
do_action( 'edit_form_advanced', $post );
$_edit_form_advanced_output = ob_get_clean();
```

We'd then write `$_edit_form_advanced_output` to the page wherever appropriate.

**Comment by moorscode** [comment]
Sounds like a good plan.

**Comment by danielbachhuber** [comment]
Related https://core.trac.wordpress.org/ticket/45283

**Comment by mtias** [comment]
I'm closing this based on https://core.trac.wordpress.org/changeset/43882 . Please reopen if I'm wrongly assuming relationship.

**Comment by garrett-eclipse** [comment]
Hi @mtias 
I confirmed that the changeset you mentioned resolved this on 5.0 RC in the case of Disable Comments as it does something similar with injecting hidden inputs. I'm wondering if that change should be ported into the plugin as well for plugin-only users and users of 5.0 that have the plugin installed as those users will still experience this issue until they adopt WP5.0 and drop Gutenberg.
Thank you
